ocaml-lsp-server/src/custom_requests/req_type_enclosing.ml

```diff
@@ -0,0 +1,224 @@
+      let loc = Merlin_parsing.Location_aux.union bl dl in
+      let txt = base ^ "." ^ reify dot in
+      Merlin_parsing.Location.mkloc txt loc
+    in
+    [ List.fold_left tail ~init:base ~f ]
+
+let overlap_with_range_end range = function
+  | None -> true
+  | Some position ->
+    let lend = range.Range.end_ in
+    if lend.line = position.Position.line then
+      lend.character > position.character
+    else lend.line > position.line
+
+let collect_all_results verbosity small_enclosing range_end result =
+  let ppf = Format.str_formatter in
+  List.filter_map
+    ~f:(fun (loc, text, _tail) ->
+      let range = Range.of_loc loc in
+      let ret x = Some (range, x) in
+      let open Merlin_analysis in
+      match text with
+      | _ when not (overlap_with_range_end range range_end) -> None
+      | Type_enclosing.String str -> ret str
+      | Type_enclosing.Type (env, t) ->
+        Type_utils.Printtyp.wrap_printing_env env ~verbosity (fun () ->
+            Type_utils.print_type_with_decl ~verbosity env ppf t);
+        ret (Format.flush_str_formatter ())
+      | Type_enclosing.Type_decl (env, id, t) ->
+        Type_utils.Printtyp.wrap_printing_env env ~verbosity (fun () ->
+            Type_utils.Printtyp.type_declaration env id ppf t);
+        ret (Format.flush_str_formatter ())
+      | Type_enclosing.Modtype (env, m) ->
+        Type_utils.Printtyp.wrap_printing_env env ~verbosity (fun () ->
+            Type_utils.Printtyp.modtype env ppf m);
+        ret (Format.flush_str_formatter ()))
+    (small_enclosing @ result)
+  |> Merlin_utils.Std.List.merge_cons
+       ~f:(fun (a_range, a_typ) (b_range, b_typ) ->
+         if Range.compare a_range b_range = Eq && String.equal a_typ b_typ then
+           Some (b_range, b_typ)
+         else None)
+
+let type_enclosing text_document_position verbosity range_end pipeline =
+  let typer = Mpipeline.typer_result pipeline in
+  let pos = get_position pipeline text_document_position in
+  let structures =
+    Mbrowse.(enclosing pos [ of_typedtree @@ Mtyper.get_typedtree typer ])
+  in
+  let path =
+    match structures with
+    | [] -> []
+    | browse -> Merlin_analysis.Browse_misc.annotate_tail_calls browse
+  in
+  let result = Merlin_analysis.Type_enclosing.from_nodes ~path in
+  let expression = reconstruct_identifier pipeline pos in
+  let small_enclosing =
+    Merlin_analysis.Type_enclosing.from_reconstructed
+      ~nodes:structures
+      ~cursor:pos
+      ~verbosity
+      expression
+  in
+  collect_all_results verbosity small_enclosing range_end result
+
+let render_result index result =
+  let current_typ =
+    match index |> List.nth result with
+    | None -> "<no information>"
+    | Some (_, typ) -> typ
+  in
+  let enclosings =
+    List.map
+      ~f:(fun (loc, typ) ->
+        `Assoc
+          [ ("start", Position.yojson_of_t loc.Range.start)
+          ; ("end_", Position.yojson_of_t loc.Range.end_)
```

Comment:
I don't think we need the `_` in the field name.

Reply:
It is the name in `Types`.